src/Native/Runtime/portable.cpp

```diff
@@ -148,7 +148,17 @@ COOP_PINVOKE_HELPER(Array *, RhpNewArray, (EEType * pArrayEEType, int numElement
     }
 
     UInt8* result = acontext->alloc_ptr;
+#ifdef HOST_64BIT
```

Comment:
It should be faster to write it like this:
```
BYTE *allocPtr = allocContext->alloc_ptr;
_ASSERTE(allocPtr <= allocContext->alloc_limit);

if ((SIZE_T)(acontext->alloc_limit - alloc_ptr) <= size)
{
    pObject = (Array *)allocPtr;
    acontext->alloc_ptr = >allocPtr + size;
    pObject->set_EEType(pArrayEEType);
    ...
}
```

It is the pattern used by CoreCLR allocation helpers: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/jithelpers.cpp . It is faster on 32-bit platforms because of 64-bit math is typically more than 2x slower than 32-bit math on 32-bit platforms. 

Also, it works for both 32-bit and 64-bit. The overflow can technically happen on 64-bit platforms too.

Reply:
Thanks, shall I do something similar for the alignment addition using `SIZE_MAX`.  Also for x64 targets I'm using 12 as the size for `g_pFreeObjectEEType` I guess that is wrong?

Reply:
> alignment addition using `SIZE_MAX`

I am not sure what you mean.

> using 12 as the size for g_pFreeObjectEEType I guess that is wrong?

Right, size of `g_pFreeObjectEEType` on 64-bit platforms is 24. The Align8 methods do not make sense and won't be used on 64-bit platforms at all, so it is better to exclude them on 64-bit platforms.

Reply:
> > alignment addition using `SIZE_MAX`
> 
> I am not sure what you mean.

For the alloc ptr we do something like `(size_t)(acontext->alloc_limit - alloc_ptr) >= paddedSize` to make sure there is space which together with the assert on `alloc_ptr <= acontext->alloc_limit` ensures its not going to overflow.  When adding the 12 bytes to the size for the alignment, it could technically overflow as well and currently that is handled by using 64 bit arithmetic.  Instead it could be `if(size > SIZE_MAX -12) throw...`

Reply:
Sounds good